service/round.go

```diff
@@ -271,13 +303,62 @@ func (r *round) calcMembersAndStatement() ([][]byte, []byte, error) {
 	return members, mtree.Root(), nil
 }
 
-func (r *round) teardown(cleanup bool) error {
+func (r *round) teardown(ctx context.Context, cleanup bool) error {
+	logger := logging.FromContext(ctx)
+	logger.Info("tearing down round", zap.String("round", r.ID), zap.Bool("cleanup", cleanup))
+	started := time.Now()
+	defer logger.Info(
+		"finished tearing down round",
+		zap.String("round", r.ID),
+		zap.Duration("duration", time.Since(started)),
+	)
+
+	if r.metricsCollector != nil {
+		prometheus.Unregister(r.metricsCollector)
+	}
+
 	if err := r.challengesDb.Close(); err != nil {
-		return err
+		return fmt.Errorf("closing DB: %w", err)
 	}
 
 	if cleanup {
 		return os.RemoveAll(r.datadir)
 	}
 	return r.saveState()
 }
+
+// Implementation of the Collector interface.
+func (r *roundMetricCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- r.totalMembersDesc
+}
+
+// Implementation of the Collector interface.
+func (r *roundMetricCollector) Collect(ch chan<- prometheus.Metric) {
+	m, err := prometheus.NewConstMetric(
+		r.totalMembersDesc,
+		prometheus.GaugeValue,
+		float64(r.round.members.Load()),
+	)
+	if err != nil {
+		logging.FromContext(context.Background()).Error("failed to create metric", zap.Error(err))
+	} else {
+		ch <- m
+	}
+}
+
+type roundMetricCollector struct {
+	round            *round
+	totalMembersDesc *prometheus.Desc
+}
+
+func newRoundsMetricCollector(round *round) prometheus.Collector {
+	return &roundMetricCollector{
+		round: round,
+		totalMembersDesc: prometheus.NewDesc(
+			prometheus.BuildFQName("poet", "round", "members_total"),
+			"the total number of members in a round",
+			nil,
+			prometheus.Labels{"epoch": round.ID},
```

Comment:
this way it doesn't report unregistered rounds?

Reply:
Why the wrapper around the collector to access round upon a request?

I think it would be simpler to create a counter similar to how the leavesCounter is set up at the moment and call Inc() on it whenever a new member successfully submitted.

This would also remove the need to make members an atomic.uint64

Reply:
To have the metric with a label you probably need CounterVec instead of Counter.

Initialize and register it when the server starts and deregister it upon shutdown.  It can then be updated by calling `memberMetric.WithLabelValues(roundID).Add(1)` in the `submit` method.

Reply:
@dshulyak 
>this way it doesn't report unregistered rounds?

Yes. The key was to have distinct `Desc`.

@fasmat 
> To have the metric with a label you probably need CounterVec instead of Counter.

As @dshulyak said, prometheus doesn't unregister the `Counter`s and it keeps growing. It grows 1 counter/2 weeks but still.

The approach with a `Collector` seems overcomplicated to me. I'm considering changing it to just a `Counter` with a const label for the epoch. Unregistering works fine (if `CounterVec` is **not** used).